design-proposals/persist-vmi-firmware-uuid.md

```diff
@@ -0,0 +1,181 @@
+# Overview
+This proposal introduces a mechanism to persist the firmware UUID of a Virtual Machine Instance (VMI) in KubeVirt.
+By storing the firmware UUID, we ensure that it remains consistent across VMI restarts.
+which is crucial for applications and services that rely on the UUID for identification or licensing purposes.
+
+## Motivation
+By definition, UUID must be a universally-unique identifier.
+The current implementation for the automatically-computed Firmware UUID is not universally unique: 
+Two VMs with the same name get the same UUID, even if they reside in different namespaces or different clusters.
+This proposal discusses how to fix this, while ensuring that existing VMs keep their current ID.
+
+## Goals
+* Generate and persist universally-unique Firmware.UUID for new VMs
+* Keep the Firmware.UUID of pre-existing VMs
+
+## Non Goals
+
+
+## Definition of Users
+End Users: Individuals or organizations running VMs and VMIs on KubeVirt who require consistent firmware UUIDs for their applications.
+
+## User Stories
+As an end-user, I expect my VMI to maintain its identity across restarts.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+
+This design provides two approaches for implementing a persistent, universally unique firmware UUID for VMs: 
+a single-phase solution with immediate impact and a two-phased approach that gradually transitions to global uniqueness.
+
+## Single-Phase Solution for Persistent, Universally Unique UUIDs
+
+### 1. Upgrade-Specific Persistence of Firmware.UUID
+
+**Description:** Prior to upgrading KubeVirt, persist the `Firmware.UUID` of existing VMs in `vm.spec`. 
+After the upgrade, any VM without a `Firmware.UUID` in its spec will be considered new.
+For these new VMs, the system defaults to using `vm.metadata.uid` as the firmware UUID if `vm.spec.template.spec.Firmware.UUID`is not defined.
+
+**Pros:**
+- The upgrade-specific code can be safely removed after two or three releases.
+- Simple logic post-upgrade, with a straightforward UUID assignment.
+- Limited disturbance to GitOps workflows, affecting only pre-existing VMs with the current method.
```

Comment:
This is only true if gitops is setting metadata.uid, correct?

Reply:
@jangel97 what is the benefit of storing this in a new annotation, instead of an already-existing API?